eclipse-theia/theia
# Open editors widget in navigator

#### What it does

Fixes: #8936 
Adds 'Opened Editors' panel into the Navigator's layout, hidden by default

![grouping-open](https://user-images.githubusercontent.com/62660714/124015206-0797f900-d9aa-11eb-920c-c0137613d522.gif)

Includes the following functionality:
- Shows all opened saveable editors
- Provides SCM, Problem, dirty, and preview state decorators
- Provides caption affix to indicate file's workspace root, as well as relative path
- Adds buttons to close individual editors as well as toolbar buttons to Close All/Save All
- Adds context menu options to tree nodes to perform actions including: Open in Terminal, Copy Path, Save, Compare, Close All, Close Others
- Tracks active editors and will also activate editors when tree nodes are selected, pins preview widget when doubleclicked
- Follows VSCode's grouping convention (i.e. GROUP 1, GROUP 2) to main area editor widgets. If widgets are placed in other shell areas, they are grouped by area (main, bottom, left, right)
- Adds tree node items for "Save All" and "Close All" on group nodes and area nodes
- Will update ordering of widgets when they are rearranged in a tabbar

Known Issues/Areas for Improvement:
- Currently this widget is only tracking NavigatableWidgets (editor widgets) to avoid too much noise, in VSCode it seems that any 'main' area widget is tracked, but this system does not align perfectly with Theia's since we can dock widgets anywhere.
- VSCode provides additional toolbar buttons including 'New Untitled File', and 'Toggle Vertical/Horizontal Layout'
- VSCode will display a badge decoration in the toolbar when there are any unsaved editors:
![image](https://user-images.githubusercontent.com/62660714/113449790-3828c080-93c4-11eb-8c7e-1561868d2eb9.png)
- VSCode auto-sizes its panel layout based on the widget's contents, I'm not quite sure how to achieve this easily with Phosphor's PanelLayout.
- When widgets are rearranged on the left/right toolbars, the ordering will not update (it seems that the left/right DockPanel's `layoutModified` listener behaves differently from the main/bottom dock panel

#### How to test

Set Up:
- Pull branch and build
- Open a single-root workspace and open some files
- Right click the file explorer's tabbar and select "Open Editors" to enable the widget

Widget Tracking:
- Open some files and observe that editors are tracked and added
- Click on the nodes and observe the their corresponding widgets are activated
- Close widgets from the tabbar, are they removed from the tree view?
- Dock widgets in different panes, split widgets in the main area, observe the grouping strategy. Does this make sense?
- Try out some of the grouping toolbar items like "Close Group" and "Save All in Group"

Decorations:
- Start editing a file, observe the the file is marked as dirty
- If a widget is opened in Preview mode, it should appear in italics (reflecting the Main Area tabbar state) double click on its node to pin the widget, it should be displayed as normal
- Attempt to close the file using the 'x' icon, does it behave as expected?
- Try these steps in a multiroot workspace, observe that caption suffixes are added that indicate the workspace root as well as its relative path
- Introducing invalid syntax in a file should decorate the node using the problem decorator to match its appearance in the file explorer
- Introducing changes will decorate the node with the SCM decorator

Actions:
- Use the toolbar icons and verify they behave as expected
- Use the context menu and verify the items behave as expected


#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
